File: DentalWave/src/main/java/com/summerproject2026/DentalWave/service/CalendarService.java
package com.summerproject2026.DentalWave.service;

import com.summerproject2026.DentalWave.dto.CalendarDto;
import com.summerproject2026.DentalWave.dto.ScheduleDto;

import java.time.LocalDate;
import java.util.List;

/**
 * Service interface defining all business operations for Calendar management.
 * Handles CRUD operations, publishing lifecycle, and nested schedule management.
 */
public interface CalendarService {

    /**
     * Creates a new calendar from the provided DTO.
     *
     * @param calendarDto the data for the new calendar
     * @return the persisted calendar as a DTO (with generated ID)
     */
    CalendarDto createCalendar(CalendarDto calendarDto);

    /**
     * Retrieves a calendar by its ID.
     *
     * @param id the calendar ID
     * @return the matching CalendarDto
     */
    CalendarDto getCalendarById(Long id);

    /**
     * Returns all calendars in the system.
     *
     * @return list of all CalendarDtos (may be empty)
     */
    List<CalendarDto> getAllCalendars();

    /**
     * Updates an existing calendar with new field values.
     *
     * @param id          the ID of the calendar to update
     * @param calendarDto the updated data
     * @return the updated CalendarDto
     */
    CalendarDto updateCalendar(Long id, CalendarDto calendarDto);

    /**
     * Deletes a calendar (and its schedules via cascade).
     *
     * @param id the ID of the calendar to delete
     */
    void deleteCalendar(Long id);

    /**
     * Marks a calendar as published, making it visible to staff.
     *
     * @param id the calendar ID
     * @return the updated CalendarDto with published = true
     */
    CalendarDto publishCalendar(Long id);

    /**
     * Reverts a calendar to draft (unpublished) state.
     *
     * @param id the calendar ID
     * @return the updated CalendarDto with published = false
     */
    CalendarDto unpublishCalendar(Long id);

    /**
     * Retrieves all calendars for a given month label.
     *
     * @param month month label to filter by (e.g. "June 2025")
     * @return list of matching CalendarDtos
     */
    List<CalendarDto> getCalendarsByMonth(String month);

    /**
     * Retrieves all calendars that have been published.
     *
     * @return list of published CalendarDtos
     */
    List<CalendarDto> getPublishedCalendars();

    /**
     * Creates a new schedule and adds it to the specified calendar.
     *
     * @param calendarId  the ID of the parent calendar
     * @param scheduleDto the data for the new schedule
     * @return the persisted ScheduleDto (with generated ID)
     */
    ScheduleDto addSchedule(Long calendarId, ScheduleDto scheduleDto);

    /**
     * Removes a schedule from a calendar and deletes it.
     *
     * @param calendarId the ID of the parent calendar
     * @param scheduleId the ID of the schedule to remove
     */
    void removeSchedule(Long calendarId, Long scheduleId);

    /**
     * Auto-generates a draft calendar for the given office and month.
     * Creates one schedule per weekday (Monday-Friday) within the date range,
     * and automatically assigns employees to teams based on role:
     * each team gets 1 Doctor + 1 TC, with Assistants distributed
     * as evenly as possible across the teams created.
     *
     * @param calendarDto the office, month, date range, and creator info
     * @return the fully populated CalendarDto, unpublished (draft state)
     */
    CalendarDto generateCalendar(CalendarDto calendarDto);

    /**
     * Schedules an employee across every day in the given calendar.
     * For each day's schedule, the employee is added to whichever
     * team currently has the fewest members, balancing team sizes.
     *
     * @param calendarId the calendar to schedule the employee into
     * @param employeeId the employee to schedule
     * @return the updated CalendarDto with the employee added to each day's smallest team
     */
    CalendarDto scheduleEmployeeAcrossCalendar(Long calendarId, Long employeeId);

    /**
     * Removes an employee from every team across every day in the
     * given calendar.
     *
     * @param calendarId the calendar to remove the employee from
     * @param employeeId the employee to remove
     * @return the updated CalendarDto with the employee removed from every team
     */
    CalendarDto removeEmployeeFromCalendar(Long calendarId, Long employeeId);

    /**
     * Removes an employee from every team on schedules that fall within
     * the given date range, across all calendars for the given office.
     * Used by Manager "Approved Request" page — "Remove from Schedule" action.
     *
     * @param officeId   the office whose calendars to search
     * @param employeeId the employee to remove
     * @param startDate  the first date of the time-off range (inclusive)
     * @param endDate    the last date of the time-off range (inclusive)
     */
    void removeEmployeeFromScheduleOnDates(Long officeId, Long employeeId,
                                           LocalDate startDate, LocalDate endDate);
}

File: DentalWave-frontend/src/services/CalendarService.js
import axios from 'axios'
import { getAuthHeader } from './AuthService'
import { apiUrl } from './apiConfig'

/**
 * Base URL for Calendar endpoints.
 */
const CALENDAR_REST_API_BASE_URL = apiUrl('/api/calendars')

/**
 * Creates a new calendar.
 *
 * @param calendar the calendar data to create
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const createCalendar = (calendar) =>
    axios.post(CALENDAR_REST_API_BASE_URL, calendar, getAuthHeader())

/**
 * Gets a single calendar by id.
 *
 * @param calendarId the calendar id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getCalendarById = (calendarId) =>
    axios.get(`${CALENDAR_REST_API_BASE_URL}/${calendarId}`, getAuthHeader())

/**
 * Gets all calendars.
 *
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getAllCalendars = () =>
    axios.get(CALENDAR_REST_API_BASE_URL, getAuthHeader())

/**
 * Updates an existing calendar.
 *
 * @param calendarId the calendar id
 * @param calendar the updated calendar data
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const updateCalendar = (calendarId, calendar) =>
    axios.put(`${CALENDAR_REST_API_BASE_URL}/${calendarId}`, calendar, getAuthHeader())

/**
 * Deletes a calendar.
 *
 * @param calendarId the calendar id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const deleteCalendar = (calendarId) =>
    axios.delete(`${CALENDAR_REST_API_BASE_URL}/${calendarId}`, getAuthHeader())

/**
 * Publishes a calendar, making it visible to staff.
 *
 * @param calendarId the calendar id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const publishCalendar = (calendarId) =>
    axios.post(`${CALENDAR_REST_API_BASE_URL}/${calendarId}/publish`, null, getAuthHeader())

/**
 * Reverts a calendar to draft state.
 *
 * @param calendarId the calendar id
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const unpublishCalendar = (calendarId) =>
    axios.post(`${CALENDAR_REST_API_BASE_URL}/${calendarId}/unpublish`, null, getAuthHeader())

/**
 * Gets all calendars for a given month label.
 *
 * @param month URL-encoded month label, e.g. "June 2026"
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getCalendarsByMonth = (month) =>
    axios.get(`${CALENDAR_REST_API_BASE_URL}/month/${encodeURIComponent(month)}`, getAuthHeader())

/**
 * Gets all published calendars.
 *
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const getPublishedCalendars = () =>
    axios.get(`${CALENDAR_REST_API_BASE_URL}/published`, getAuthHeader())

/**
 * Adds a new schedule to a calendar.
 *
 * @param calendarId the parent calendar id
 * @param schedule the schedule data to add
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const addScheduleToCalendar = (calendarId, schedule) =>
    axios.post(`${CALENDAR_REST_API_BASE_URL}/${calendarId}/schedules`, schedule, getAuthHeader())

/**
 * Removes a schedule from a calendar.
 *
 * @param calendarId the parent calendar id
 * @param scheduleId the schedule id to remove
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const removeScheduleFromCalendar = (calendarId, scheduleId) =>
    axios.delete(
        `${CALENDAR_REST_API_BASE_URL}/${calendarId}/schedules/${scheduleId}`,
        getAuthHeader()
    )

/**
 * Auto-generates a draft calendar for the given office and month.
 * Creates one schedule per weekday with employees automatically
 * assigned to teams based on role (Doctor + TC + Assistants).
 *
 * @param calendar the office, month, date range, and creator info
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const generateCalendar = (calendar) =>
    axios.post(`${CALENDAR_REST_API_BASE_URL}/generate`, calendar, getAuthHeader())

/**
 * Schedules an employee across every day in the given calendar.
 * The employee is added to the smallest team on each day.
 *
 * @param calendarId the calendar to schedule the employee into
 * @param employeeId the employee to schedule
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const scheduleEmployeeAcrossCalendar = (calendarId, employeeId) =>
    axios.post(
        `${CALENDAR_REST_API_BASE_URL}/${calendarId}/employees/${employeeId}/schedule-all`,
        null,
        getAuthHeader()
    )

/**
 * Removes an employee from every team across every day in the calendar.
 *
 * @param calendarId the calendar to remove the employee from
 * @param employeeId the employee to remove
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const removeEmployeeFromCalendar = (calendarId, employeeId) =>
    axios.delete(
        `${CALENDAR_REST_API_BASE_URL}/${calendarId}/employees/${employeeId}/schedule-all`,
        getAuthHeader()
    )

/**
 * Removes an employee from every team on schedules within a date range
 * for the given office. Used by Manager "Approved Request" page.
 *
 * @param officeId the office whose calendars to search
 * @param employeeId the employee to remove
 * @param startDate ISO date string e.g. '2026-06-15'
 * @param endDate ISO date string e.g. '2026-06-20'
 * @returns {Promise<axios.AxiosResponse<any>>}
 */
export const removeEmployeeFromScheduleOnDates = (officeId, employeeId, startDate, endDate) =>
    axios.delete(
        `${CALENDAR_REST_API_BASE_URL}/office/${officeId}/employees/${employeeId}/dates?startDate=${startDate}&endDate=${endDate}`,
        getAuthHeader()
    )
